accepted/00000-modular-repos-with-clm.md

```diff
@@ -0,0 +1,179 @@
+ - Enabled
+ - Disabled
+ - Neither enabled nor disabled
+
+When a module is neither enabled nor disabled, all the alternative streams and contexts of a stream are available to the package manager. This state confuses Uyuni when comparing the package versions.
+
+The key to building a regular repository with no conflicting packages is to simulate the modules in a way that every single module is either "enabled" or "disabled".
+
+Therefore, the high-level algorithm is proposed as:
+ 1. Let the users pick desired streams so that every module is either "enabled" or "disabled"
+ 2. Compile a blacklist of RPM artifacts from the "disabled" modules
+ 3. Generate a "regular" repository, filtering out the blacklisted RPMs
+
+The design utilizes content lifecycle management filters and repository building capabilities to implement the algorithm above.
+
+## Components
+
+This section summarizes the different components involved in the design.
+
+### Backend
+The Java backend is responsible for filtering out unselected module streams and building the repository with the desired packages through CLM.
+
+### API
+A Python API parses and interprets the module metadata, resolves modular dependencies and translates specified modules and streams to an actual package list.
+The API utilizes the Python port of [libmodulemd](https://github.com/fedora-modularity/libmodulemd), which is used for structural parsing of `modules.yaml` files.
+
+An example of such implementation is provided in [attachments/00000-modular-repos-api.py](attachments/00000-modular-repos-api.py).
+
+### UI
+The UI consists of a new type of CLM filter called "module". The filter provides inputs for a module name and a stream name.
+Each filter selects a stream for a single module. The stream input can be left empty to select the default stream.
+Since a CLM filter is not related to any channels, it is not possible to display a list of modules/streams on the filter form and the inputs can only be simple text inputs.
+The module information for a modular repository is displayed under an additional "Modules" tab on the "Channel Details" page.
+A modular repository is indicated with an icon next to the name of the repository in the CLM sources list. A click on the icon navigates to the "Modules" tab of the channel details.
+
+## The Process
+
+This section describes the detailed design for the different phases of a CLM project.
+
+### Project setup
+At the project setup, the backend checks if any of the sources are modular repositories.
+This check is done by doing a `getChannel().getModule()` call on the source object.
+For every modular source, a proper indicator is displayed as described in the UI component section.
+
+Currently, every CLM filter displays a field to choose one of two hard-coded rules: `allow` or `deny`. For module filters, this field must be hidden.
+Each added module filter stores the value of a chosen module/stream in `name` or `name:stream` format.
```

Comment:
Not sure if this would work, as a customer can add conflicting filters.
Additionally a created filter does not mean that that filter gets assigned to a project.

I think we need to find a way to have only 1 filter per project and you set everything in this one filter.

Reply:
This is all about the UX. Choosing either way is just a matter of designing the UI accordingly, and serializing the filter data to a string. I guess both are achievable. But since the filter forms are modal dialogs, I wonder if it would be convenient to list all modules in one.

> Not sure if this would work, as a customer can add conflicting filters.

On the other hand, a part of the implementation will be to give user early feedback with red notification errors as soon as they input an invalid combination, I don't think it's a problem.
We can do a sanity check whenever an individual filter gets updated in real-time to check if the update plays well with the existing filters and sources and notify the user.

Reply:
> We can do a sanity check whenever an individual filter gets updated in real-time to check if the update plays well with the existing filters and sources and notify the user.

I'll add a part to explain this in detail

Reply:
I think we can add the warning for limiting to one filter only of this specific type when attaching filters to a project. Then, doing all the module filtering for a modular repository in one filter only, it can be added on top of my previous comment about a select box.
Let's split the filter in two fields: a list of all possible modular repositories, and a list of modules for the selected modular repository. Pretty the same we do in ActivationKeys selecting the parent channel and changing dynamically the list of options for child channels. Would this work?